api/api.go

```diff
@@ -947,29 +947,58 @@ func (l *VolumeLocator) MergeVolumeSpecLabels(s *VolumeSpec) *VolumeLocator {
 	return l
 }
 
-func (v *Volume) IsPermitted(ctx context.Context) bool {
-	return v.GetSpec().IsPermitted(ctx)
+func (v *Volume) IsPermitted(ctx context.Context, accessType Ownership_AccessType) bool {
+	return v.GetSpec().IsPermitted(ctx, accessType)
 }
 
-func (v *VolumeSpec) IsPermitted(ctx context.Context) bool {
-	if v.GetOwnership() != nil {
-		if userinfo, ok := auth.NewUserInfoFromContext(ctx); ok {
-			// Check Access
-			return v.IsPermittedFromUserInfo(userinfo)
-		} else {
-			// There is no user information in the context so
-			// authorization is not running
-			return true
-		}
+func (v *VolumeSpec) IsPermitted(ctx context.Context, accessType Ownership_AccessType) bool {
+	if v.IsPublic() {
+		return true
 	}
 
-	// There is no ownership on this volume, so allow access
-	return true
+	// Volume is not public, check permission
+	if userinfo, ok := auth.NewUserInfoFromContext(ctx); ok {
+		// Check Access
+		return v.IsPermittedFromUserInfo(userinfo, accessType)
+	} else {
+		// There is no user information in the context so
+		// authorization is not running
+		return true
+	}
 }
 
-func (v *VolumeSpec) IsPermittedFromUserInfo(user *auth.UserInfo) bool {
+func (v *VolumeSpec) IsPermittedFromUserInfo(user *auth.UserInfo, accessType Ownership_AccessType) bool {
+	if v.IsPublic() {
+		return true
+	}
+
 	if v.GetOwnership() != nil {
-		return v.GetOwnership().IsPermitted(user)
+		return v.GetOwnership().IsPermitted(user, accessType)
 	}
 	return true
 }
+
+func (v *VolumeSpec) IsPublic() bool {
+	return v.GetOwnership() == nil || v.GetOwnership().IsPublic()
+}
+
+// GetSnapshotCreatorOwnership returns the appropriate ownership for the
+// new snapshot and if an update is required
+func (v *VolumeSpec) GetCloneCreatorOwnership(ctx context.Context) (*Ownership, bool) {
```

Comment:
nit: should read GetCloneCreatorOwnership returns...

Reply:
doh! 👍